src/io/flutter/preview/WidgetEditToolbar.java

```diff
@@ -0,0 +1,396 @@
+      ApplicationManager.getApplication().executeOnPooledThread(() -> {
+        final OutlineOffsetConverter converter = new OutlineOffsetConverter(project, activeFile.getValue());
+        final FlutterOutline firstOutline = outlines.get(0);
+        final FlutterOutline lastOutline = outlines.get(outlines.size() - 1);
+        final int offset = converter.getConvertedOutlineOffset(firstOutline);
+        final int length = converter.getConvertedOutlineEnd(lastOutline) - offset;
+        final List<SourceChange> changes = flutterAnalysisServer.edit_getAssists(selectionFile, offset, length);
+
+        // If the current file or outline are different, ignore the changes.
+        // We will eventually get new changes.
+        final List<FlutterOutline> newOutlines = activeOutlines.getValue();
+        if (!Objects.equals(activeFile.getValue(), selectionFile) || !outlines.equals(newOutlines)) {
+          return;
+        }
+
+        // Associate changes with actions.
+        // Actions will be enabled / disabled in background.
+        for (SourceChange change : changes) {
+          final AnAction action = messageToActionMap.get(change.getMessage());
+          if (action != null) {
+            actionToChangeMap.put(action, change);
+          }
+        }
+
+        // Update actions immediately.
+        if (getToolbar() != null) {
+          ApplicationManager.getApplication().invokeLater(() -> getToolbar().updateActionsImmediately());
+        }
+      });
+    }
+  }
+
+  FlutterOutline getWidgetOutline() {
+    final List<FlutterOutline> outlines = activeOutlines.getValue();
+    if (outlines.size() == 1) {
+      final FlutterOutline outline = outlines.get(0);
+      if (outline.getDartElement() == null) {
+        return outline;
+      }
+    }
+    return null;
+  }
+
+  private void sendAnalyticEvent(@NotNull String name) {
+    FlutterInitializer.getAnalytics().sendEvent("preview", name);
+  }
+
+  private void applyChangeAndShowException(SourceChange change) {
+    ApplicationManager.getApplication().runWriteAction(() -> {
+      try {
+        AssistUtils.applySourceChange(project, change, false);
+        if (hotReloadOnAction) {
+          final InspectorService inspectorService = InspectorGroupManagerService.getInstance(project).getInspectorService();
+
+          if (inspectorService != null) {
+            ArrayList<FlutterApp> apps = new ArrayList<>();
+            apps.add(inspectorService.getApp());
+            FlutterReloadManager.getInstance(project).saveAllAndReloadAll(apps, "Refactor widget");
+          }
+        }
+      }
+      catch (DartSourceEditException exception) {
+        FlutterMessages.showError("Error applying change", exception.getMessage());
+      }
+    });
+  }
+
+  public void createPopupMenu(Component comp, int x, int y) {
+    // The corresponding tree item may have just been selected.
+    // Wait short time for receiving assists from the server.
```

Comment:
Was this existing code? We wait up to 100ms in 5ms increments?

Reply:
Yep. This is completely unchanged from the existing code and example of why we need to do a real audit of all the times we are locking the UI thread from the IntelliJ plugin.
Good news is this case is only triggered on right click in the outline view so it won't cause UI hangs any other times.